### ISSUE TITLE
Appending to support shape aspects / support shape aspects from revit

### ISSUE THREAD

**Issue by theoryshaw** [IfcOpenShell/IfcOpenShell#5839]
A type doesn't retain it's `IFCSHAPEASPECT` assignment when imported into a project. 

file: 
[Wall with Windows - Copy.ifc.txt](https://github.com/user-attachments/files/18001933/Wall.with.Windows.-.Copy.ifc.txt)

https://github.com/user-attachments/assets/ce6cf96b-4bca-49bc-93d1-19330e5c6072

**Comment by Moult** [comment]
OK let's start by saying shape aspects are really, really nasty in IFC. They're difficult and cumbersome to work with in the schema.

Yes, it's a bug where we forgot to consider shape aspects when appending from a project library. That's fixed.

However there's another bug too: our UI showed shape aspects for the window type in the Revit IFC, when it probably should not have. That's now fixed (by my interpretation), but it does mean there is either a bug in Revit or a bug in IFC or both. Ping @AngelVelezSosa @TLiebich @aothms @Andrej730 this is a subtle one.

There are 2 informal propositions in the docs for shape aspect:

> 1. If ShapeRepresentations points to shape representations that are part of an IfcProductDefinitionShape, PartOfProductDefinitionShape must refer to this instance of IfcProductDefinitionShape.
> 2. If ShapeRepresentations points to shape representations that are part of an IfcRepresentationMap, PartOfProductDefinitionShape must refer to this instance of IfcRepresentationMap.

Here is the diagram taken from the docs on shape aspects. Notice the propositions aren't exactly accurate. Because the `ShapeRepresentations` in the diagram points to an `IfcShapeRepresentation` that is completely independent. It's actually the `IfcFacetedBrep` which is shared (and indirectly related back to an IfcProduct/IfcTypeProduct).

![Image](https://github.com/user-attachments/assets/f00a9919-eaf2-4aed-9e0e-33d010811d2a)

I've added another diagram next to it, describing the (implied, but never explicitly drawn) situation for types. I've also added a third diagram below it combining the Mapped Geometry concept template (showing how occurrences map representations from their types) and what I expect to happen in that scenario.

Notice how in the bottom diagram ... it's not possible to satisfy _both_ informal propositions in the docs. The `IfcFacetedBrep` is equally both part of an `IfcRepresentationMap` as well as an `IfcProductDefinitionShape`.

In that third case, I propose that it's more correct to have the `IfcShapeAspect` point to the `IfcRepresentationMap` because that is shared across both types and occurrences. It also means that when a type is part of a project library (like in the original bugreport) and exists in isolation without any occurrences, the shape aspects are not lost.

Anyway. There it is. Here's a modified diagram showing what Revit does. Revit (2024) seems to always link shape aspects back to the product's `IfcProductDefinitionShape` and never to the `IfcRepresentationMap`. It's not exactly "wrong" since the IFC docs are in bad shape in this area, but it's not exactly right either by my analysis :)

![Image](https://github.com/user-attachments/assets/33bb927b-f9c0-4008-bf4c-451ffef09b73)

**Comment by AngelVelezSosa** [comment]
These things make my head hurt - interested to hear from Thomas L and also Evandro as to how this will be dealt with in the Validation Service.

**Comment by Andrej730** [comment]
Makes sense to me, just connecting to Representation Map seems to be the most straightforward way to do it.

In terms of IFC specs even switching order of formal propositions would help - so relation to IfcRepresentationMap would take the priority.

> Anyway. There it is. Here's a modified diagram showing what Revit does. Revit (2024) seems to always link shape aspects back to the product's `IfcProductDefinitionShape` and never to the `IfcRepresentationMap`. It's not exactly "wrong" since the IFC docs are in bad shape in this area, but it's not exactly right either by my analysis :)

What Revit does if there are 2 occurrences sharing a representation - it creates 2 IfcShapeAspects each one having it's own IfcShapeRepresentation?
